server/vertx/src/main/scala/sttp/tapir/server/vertx/decoders/VertxInputDecoders.scala

```diff
@@ -0,0 +1,128 @@
+    decodeBodyAndInputs(endpoint, rc) match {
+      case values: DecodeInputsResult.Values =>
+        InputValues(endpoint.input, values) match {
+          case InputValuesResult.Value(params, _) =>
+            logicHandler(params)
+          case InputValuesResult.Failure(_, failure) =>
+            tryEncodeError(endpoint, rc, failure)
+        }
+      case DecodeInputsResult.Failure(input, failure) =>
+        val decodeFailureCtx = DecodeFailureContext(input, failure)
+        endpointOptions.decodeFailureHandler(decodeFailureCtx) match {
+          case DecodeFailureHandling.NoMatch =>
+            endpointOptions.logRequestHandling.decodeFailureNotHandled(endpoint, decodeFailureCtx)(endpointOptions.logger)
+            rc.response.setStatusCode(404).end()
+          case DecodeFailureHandling.RespondWithResponse(output, value) =>
+            endpointOptions.logRequestHandling.decodeFailureHandled(endpoint, decodeFailureCtx, value)(endpointOptions.logger)
+            VertxOutputEncoders.apply(output, value)(endpointOptions)(rc)
+        }
+
+    }
+  }
+
+  private def decodeBodyAndInputs(e: Endpoint[_, _, _, _], rc: RoutingContext)
+                                         (implicit serverOptions: VertxEndpointOptions): DecodeInputsResult =
+    decodeBody(DecodeInputs(e.input, new VertxDecodeInputsContext(rc)), rc)
+
+  private def decodeBody(result: DecodeInputsResult, rc: RoutingContext)
+                        (implicit serverOptions: VertxEndpointOptions): DecodeInputsResult = {
+    result match {
+      case values: DecodeInputsResult.Values =>
+        values.bodyInput match {
+          case None => values
+          case Some(bodyInput@EndpointIO.Body(bodyType, codec, _)) =>
+            codec.decode(extractRawBody(bodyType, rc)) match {
+              case DecodeResult.Value(body) => values.setBodyInputValue(body)
+              case failure: DecodeResult.Failure => DecodeInputsResult.Failure(bodyInput, failure): DecodeInputsResult
+            }
+        }
+      case failure: DecodeInputsResult.Failure => failure
+    }
+  }
+
+  private def extractRawBody[B](bodyType: RawBodyType[B], rc: RoutingContext)
+                       (implicit serverOptions: VertxEndpointOptions): Any =
+    bodyType match {
+      case RawBodyType.StringBody(defaultCharset) => rc.getBodyAsString(defaultCharset.toString).get
+      case RawBodyType.ByteArrayBody => rc.getBodyAsString.get.getBytes()
+      case RawBodyType.ByteBufferBody => rc.getBody.get.getByteBuf.nioBuffer()
+      case RawBodyType.InputStreamBody =>
+        new ByteArrayInputStream(rc.getBody.get.getBytes) // README: be really careful with that
+      case RawBodyType.FileBody =>
+        rc.fileUploads().toList match {
+          case List(upload) =>
+            new File(upload.uploadedFileName())
+          case List() if rc.getBody.isDefined => // README: really weird, but there's a test that sends the body as String, and expects a File
```

Comment:
you never know, you might require a plain body (from the perspective of the client) to be saved to a file (on the server) :)

Reply:
Sure, the only problem I found with this, is that creating a file and writing to it is an asynchronous operation by essence, but ` implicit case object FileBody extends Binary[File]` therefore I had to use Vert.x's blocking methods in order to write to the file.
It should be (at least in Vert.x context) a  `Future[File]]`. Here, the implementation could lead to dramatic performance issues (blocking the event-loop).

Reply:
Hm not sure I understand - what would be the difference between consuming a `File` and `Future[File]`? Both in the end would require using that file (path) to write the content, synchronously or asynchronously?

A strict `File` can be easily converted to a future using `Future.successful(file)` :)

Reply:
The issue is simply a lack of knowledge on my part.

Since ` FileBody`  extends ` Binary[File]` am I allowed to return ` Future[File]]` ?

Reply:
yes, sure, that's just an indication of how to handle the body. It can be handled asynchronously - the whole seerver logic returns a `Future` anyway, so if any of your intermediate methods are async, you can just `flatMap` and combine the futures.

Reply:
Mmh, I got a classcastexception. I'll investigate.

Reply:
yes, you're changing the signature then :) You can wrap any strict values with `Future.successful`, though

Reply:
Done. Thank you.